plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol

```diff
@@ -0,0 +1,224 @@
+                    withdrawFromVault(self, withdrawal);
+                    withdrawal.exitTarget.transfer(withdrawal.piggybackBondSize);
+                    emit InFlightExitOutputWithdrawn(exitId, i);
+                }
+            }
+        }
+
+        flagInputAndOutputSpent(self.framework, exit, token);
+        clearPiggybackFlag(exit, token);
+
+        if (allPiggybackCleared(exit)) {
+            exit.bondOwner.transfer(exit.bondSize);
+            delete exitMap.exits[exitId];
+        }
+    }
+
+    function isAnyInputSpent(
+        PlasmaFramework framework,
+        PaymentExitDataModel.InFlightExit memory exit,
+        address token
+    )
+        private
+        view
+        returns (bool)
+    {
+        uint256 inputNumOfTheToken;
+        for (uint i = 0; i < MAX_INPUT_NUM; i++) {
+            if (exit.inputs[i].token == token) {
+                inputNumOfTheToken++;
+            }
+        }
+        bytes32[] memory outputIdsOfInputs = new bytes32[](inputNumOfTheToken);
+        uint sameTokenIndex = 0;
+        for (uint i = 0; i < MAX_INPUT_NUM; i++) {
+            if (exit.inputs[i].token == token) {
+                outputIdsOfInputs[sameTokenIndex] = exit.inputs[i].outputId;
+                sameTokenIndex++;
+            }
+        }
+        return framework.isAnyOutputsSpent(outputIdsOfInputs);
+    }
+
+    function shouldWithdrawInput(
+        PaymentExitDataModel.InFlightExit memory exit,
+        PaymentExitDataModel.WithdrawData memory withdrawal,
+        address token,
+        uint16 index
+    )
+        private
+        pure
+        returns (bool)
+    {
+        return withdrawal.token == token &&
+                exit.isInputPiggybacked(index);
+    }
+
+    function shouldWithdrawOutput(
+        Controller memory controller,
+        PaymentExitDataModel.InFlightExit memory exit,
+        PaymentExitDataModel.WithdrawData memory withdrawal,
+        address token,
+        uint16 index
+    )
+        private
+        view
+        returns (bool)
+    {
+        return withdrawal.token == token &&
+                exit.isOutputPiggybacked(index) &&
+                !controller.framework.isOutputSpent(withdrawal.outputId);
+    }
+
+    function withdrawFromVault(
+        Controller memory self,
+        PaymentExitDataModel.WithdrawData memory withdrawal
+    )
+        private
+    {
+        if (withdrawal.token == address(0)) {
+            self.ethVault.withdraw(withdrawal.exitTarget, withdrawal.amount);
+        } else {
+            self.erc20Vault.withdraw(withdrawal.exitTarget, withdrawal.token, withdrawal.amount);
+        }
+    }
+
+    function flagInputAndOutputSpent(
+        PlasmaFramework framework,
+        PaymentExitDataModel.InFlightExit memory exit,
+        address token
+    )
+        private
+    {
+        uint256 inputNumOfTheToken;
+        for (uint16 i = 0; i < MAX_INPUT_NUM; i++) {
+            if (exit.inputs[i].token == token) {
```

Comment:
no, we want to mark *all* inputs as spent no matter piggybacked or not. In the case of exiting from output (canonical), all inputs cannot be used anymore. 

In the case of non-canonical, though we can argue we don't need to flag those non-piggybacked inputs. But since the original design in `RootChain.sol` does not allow re-start-IFE, so practically it means once processed, no inputs can exit anymore anyway.

However, if people think for the case of non-canonical, we'd like to only process the piggybacked inputs, I can adjust to that.

(btw, I will add comment to this)